### ISSUE TITLE
Multiple output options simultaneously

### ISSUE THREAD

**Issue by ben-elttam** [semgrep/semgrep#3019]
**Is your feature request related to a problem? Please describe.**

Often I like to get the default text log output of `semgrep`, but also re-run with `--json` and `--sarif` to save these options as well.
I do want to have to run `semgrep` 3 separate times with slightly different logging options.

**Describe the solution you'd like**

I would like to be able to execute `semgrep` once and get all of these outputs, something like `semgrep ... --log adsf.log --json asdf.json --sarif asdf.sarif.json`

**Describe alternatives you've considered**

Run `semgrep` 3 separate times, e.g. `semgrep > asdf.log; semgrep --json > asdf.json; semgrep --sarif > asdf.sarif.json`

**Additional context**

The popular tool `nmap` has an option `-oA` to save all the different types of logging at once (e.g. nmap (text), xml, and greppable). This isn't strictly necessary but could be considered if r2c likes it.

I often run semgrep via docker, saving multiple file types would require mounting a volume, or running separately if streaming (otherwise would need to stream a tar or similar which is too complicated IMO).

I do actually like to have  the colour output as I can handle Ansi Escape sequences in my favourite editor, but a syntax file is probably a better solution.

**Comment by mschwager** [comment]
Hey @ben-elttam,

What's your sense of priority on this? This sounds like a nice-to-have feature request to me.

**Comment by ben-elttam** [comment]
Yeah it is a nice to have. Not that urgent. Just means saving multiple identical passes to get different log outputs.

**Comment by emjin** [comment]
Related (duplicate?): https://github.com/returntocorp/semgrep/issues/3699. Opened by product; idk if that means priority is getting bumped

**Comment by gabpil45** [comment]
Hi, Can we reopen this? We would like to use it to utilize the JUnit output to post it to Jenkins and the Sarif/JSON to the github API.

Thanks,
G

**Comment by agonzalez-plume** [comment]
As posted in slack, we may be able to change OutputHandler to be an array of handlers and each time walk the array on an output operation.  Not sure of side effects, but seems possible to work. At this point just documenting an option without submitting a PR.

I can see this change happening in scan.py and semgrep_main.py

scan.py:
change output_settings to build array of settings for each type. Right now it's just one.

           output_format_array = []
           
             output_format_array.append(OutputFormat.TEXT)
               if json or json_time or debugging_json:
                   output_format_array.append(OutputFormat.JSON)
               elif junit_xml:
                   output_format_array.append(OutputFormat.JUNIT_XML)
               elif sarif:
                   output_format.array.append)OutputFormat.SARIF)
               elif emacs:
                   output_forma_array.append(OutputFormat.EMACS)
               elif vim:
                   output_format_array.append(OutputFormat.VIM)
           
           output_handler_array=[]
           for output_format in output_format_array:
               output_settings = OutputSettings(
                   output_format=output_format,
                   output_destination=output,
                   error_on_findings=error_on_findings,
                   strict=strict,
                   debug=debugging_json,
                   verbose_errors=verbose,
                   timeout_threshold=timeout_threshold,
                   json_stats=json_stats,
                   output_time=output_time,
                   output_per_finding_max_lines_limit=max_lines_per_finding,
                   output_per_line_max_chars_limit=max_chars_per_line,
               )
              output_handler_array.append(OutputHandler(output_settings))
           )


Then each call for using the output handler use the array. Also have to change methods to take array for output handler.

           except SemgrepError as e:
                for output_handler in output_handler_array: 
                  output_handler.handle_semgrep_errors([e])
                  output_handler.output({}, all_targets=set(), filtered_rules=[])
               raise e

            for output_handler in output_handler_array:
                  output_handler.output(
                      filtered_matches_by_rule,
                      all_targets=all_targets,
                      ignore_log=ignore_log,
                      profiler=profiler,
                      filtered_rules=filtered_rules,
                      profiling_data=profiling_data,
                      severities=shown_severities,
                )

**Comment by underyx** [comment]
So a nice CLI design for this would probably be:

`semgrep scan --config=auto --json one.json --sarif two.sarif`

which would be equivalent to

```
semgrep scan --config=auto --json --output one.json
semgrep scan --config=auto --sarif --output two.sarif
```

Here are some requirements for the purposes of backwards-compatibility:

1. setting `--json --output one.json` should be translated to `--json=one.json`
2. setting just `--json` should be translated to `--json=-` so that we print JSON to stdout
3. if any format requests to print to stdout, do not print text output to stdout
4. do not allow multiple formats printing to stdout in one invocation
5. **optionally**, if all requested formats are written to a file, we could still print text output to stdout?
5. **optionally**, support multiple of the same format like `--json one.json --json two.json`

**Comment by underyx** [comment]
Update on the above proposal: this would would break usages like `$ semgrep scan --json /code`. Basically, we cannot know if the token after a format flag was intended to be a scan target, or the output file path. Today this command would "print findings on /code to stdout in JSON format" but after this feature it'd "write findings in working directory to a file named /code".

@chadac pointed out this hacky logic could make it work, kinda:
```
for each --output_format:
  if value not a valid filesystem path:
    --output_format is a flag
  elif value is a directory:
    value is a target
  elif value is a nonexistent file:
    value is an output destination
  elif value is a file that exists:
    open value as a file, if it is in output_format and follows semgrep's schema, then it's an output destination
    else it's a target, throw a deprecation warning for now. In the future, make this an I/O prompt
```
which would be on top of this flag preprocessing logic
```
if there's an --output flag:
    if there's exactly one format flag:
        move the --output value to the format flag
        remove the --output flag
        print a deprecation warning
    else:
        error

if there's more than one format flag with no value:
    error

if there's exactly one format flag with no value:
    set "-" as its value
    print a deprecation warning
```
and this is a 99% correct, but very messy solution, especially cause we'd need to write code to detect if a file contains text that matches the selected output format. I'm gonna say the proposal v1 from above is unviable.

**Comment by underyx** [comment]
## Proposal v2

```
semgrep scan --config=auto --output-json one.json --output-sarif two.sarif
```
or
```
semgrep scan --config=auto --write json one.json --write sarif two.sarif
```

which would be equivalent to

```
semgrep scan --config=auto --json --output one.json
semgrep scan --config=auto --sarif --output two.sarif
```

#### Note that

- setting `--json --output one.json` should be translated to `--output-json=one.json`
- setting just `--json` should be translated to `--output-json=-`
- if any flag requests to print to stdout (`--output-foo=-`), do not print text output to stdout
- do not allow multiple formats printing to stdout in one invocation (`--output-foo=- --output-bar=-`)
- optionally, if all requested formats are written to a file, we could still print text output to stdout?
- optionally, support multiple of the same format like --json one.json --json two.json

#### Deprecations (or lack thereof)

I think we shouldn't deprecate `semgrep --json`, as it's a lot more ergonomic on the command line than `semgrep --output-json=-`.

`--output` will not be useful anymore after this, so we could deprecate that.

#### Alternatives

- `--json` could be changed to `--format=json` specifically for stdout formatting
- could have some DSL for defining outputs such as `--write json:foo.json,sarif:bar.sarif`  (yuck tho)

**Comment by ben-elttam** [comment]
anchore/syft does [multiple outputs](https://github.com/anchore/syft#multiple-outputs) this way:

Syft can also output _multiple_ files in differing formats by appending
`=<file>` to the option, for example to output Syft JSON and SPDX JSON:

```shell
syft <image> -o json=sbom.syft.json -o spdx-json=sbom.spdx.json
```

**Comment by jonny64** [comment]
could you please give anyworkaround to convert --json to human readable format ? `write a tool yourself` ?
use case:
* gitlab
* upload json results to defect portal
* still should be able to view pipeline output as a human

**Comment by elb-slack** [comment]
At Slack, we'd love to be able to have output in both semgrep's json format and in sarif. Our tooling relies on the json format (we remove false positives and do comparisons between branches or over time), but we'd like to upload SARIF to aggregation tools for additional analysis. Preferably I'd like to not run semgrep twice just to get different outputs.

Otherwise, I'm going to have to rewrite a significant portion of our tooling to support SARIF. 😓

**Comment by morriswinkler** [comment]
+1

**Comment by tarihub** [comment]
+1

**Comment by OffBy0x01** [comment]
I came to the same conclusion as has been mentioned many a time here:
> allow users to specify a file for each output as an additional argument, and if no file path is provided, the existing behaviour could be used.  `semgrep ci --json=myJsonOutput.json --gitlab-sast=gitlab-sast.json` 

The JSON or SARIF output is relatively easy to write OPA policies for, allowing users to write complex local policies in addition to what is provided by the rule board. For example: "Block pipeline if any high or medium confidence, high severity, security findings. or "block pipeline if 5 or more high severity security findings".

Currently, I am outputting as SARIF, then converting to the gitlab-sast format. Just in case anyone has a similar use-case (OPA) - gitlab has an 'internal reporting tool' that is publicly available & has a permissive license, that can convert SARIF to gitlab-sast.
https://gitlab.com/gitlab-org/security-products/analyzers/report/-/blob/main/sarif.go 

Not a perfect solution as it doesn't convert partial scans and there are a few longstanding PRs/issues - would much rather see multiple outputs supported in semgrep.

**Comment by bojackhorseman0309** [comment]
This is fixed in [1.70.0](https://github.com/semgrep/semgrep/blob/develop/CHANGELOG.md#1700---2024-04-24)
